src/rgw/rgw_rest_role.cc

```diff
@@ -551,96 +493,95 @@ void RGWModifyRoleTrustPolicy::execute(optional_yield y)
 {
-  op_ret = get_params();
-  if (op_ret < 0) {
-    return;
+  rgw::sal::RoleList listing;
+  if (!account_id.empty()) {
+    // list roles from the account
+    op_ret = driver->list_account_roles(this, y, account_id, path_prefix,
+                                        marker, max_items, listing);
+  } else {
+    // list roles from the tenant
+    op_ret = driver->list_roles(this, y, s->auth.identity->get_tenant(),
+                                path_prefix, marker, max_items, listing);
   }
-  vector<std::unique_ptr<rgw::sal::RGWRole>> result;
-  op_ret = driver->get_roles(s, y, path_prefix, s->user->get_tenant(), result);
 
   if (op_ret == 0) {
-    s->formatter->open_array_section("ListRolesResponse");
-    s->formatter->open_array_section("ListRolesResult");
-    s->formatter->open_object_section("Roles");
-    for (const auto& it : result) {
-      s->formatter->open_object_section("member");
-      it->dump(s->formatter);
-      s->formatter->close_section();
+    s->formatter->open_object_section("ListRolesResponse");
+    s->formatter->open_object_section("ListRolesResult");
+    s->formatter->open_array_section("Roles");
+    for (const auto& info : listing.roles) {
+      encode_json("member", info, s->formatter);
     }
-    s->formatter->close_section();
-    s->formatter->close_section();
+    s->formatter->close_section(); // Roles
+
+    const bool truncated = !listing.next_marker.empty();
+    encode_json("IsTruncated", truncated, s->formatter);
+    if (truncated) {
+      encode_json("Marker", listing.next_marker, s->formatter);
+    }
+
+    s->formatter->close_section(); // ListRolesResult
     s->formatter->open_object_section("ResponseMetadata");
     s->formatter->dump_string("RequestId", s->trans_id);
-    s->formatter->close_section();
-    s->formatter->close_section();
+    s->formatter->close_section(); // ResponseMetadata
+    s->formatter->close_section(); // ListRolesResponse
   }
 }
 
-int RGWPutRolePolicy::get_params()
+int RGWPutRolePolicy::init_processing(optional_yield y)
 {
   role_name = s->info.args.get("RoleName");
+  if (!validate_iam_role_name(role_name, s->err.message)) {
+    return -EINVAL;
+  }
+
   policy_name = s->info.args.get("PolicyName");
   perm_policy = s->info.args.get("PolicyDocument");
 
-  if (role_name.empty() || policy_name.empty() || perm_policy.empty()) {
-    ldpp_dout(this, 20) << "ERROR: One of role name, policy name or perm policy is empty"<< dendl;
+  if (policy_name.empty()) {
+    s->err.message = "Missing required element PolicyName";
+    return -EINVAL;
+  }
+  if (perm_policy.empty()) {
+    s->err.message = "Missing required element PolicyDocument";
     return -EINVAL;
   }
-  bufferlist bl = bufferlist::static_from_string(perm_policy);
   try {
     const rgw::IAM::Policy p(
-      s->cct, s->user->get_tenant(), bl,
+      s->cct, s->user->get_tenant(), perm_policy,
       s->cct->_conf.get_val<bool>("rgw_policy_reject_invalid_principals"));
```

Comment:
the policy verification that we do in rgw was tied to the `tenant` till now. with the `account` changes, shouldn't this be changed to `account`?

Reply:
thanks @pritha-srivastava, i _just_ ran into problems here with the new s3test case `test_account_role_policy_allow_get_role()` for "test that identity policy can grant non-root users access to iam account apis"

because i passed in an empty tenant name there, the [// You can't specify resources for someone ELSE'S account](https://github.com/ceph/ceph/blob/ed736fab/src/rgw/rgw_iam_policy.cc#L628-L634) block causes us to parse `'Resource': '*'` as `Resource: [ arn:*:*:*::* ]` instead of `arn:*:*:*:*:*` so the comparison fails against role arns like `arn:aws:iam::RGW11111111111111111:role/s3-tests/s3-tests-MyRole`. i shared log snippet for this in https://gist.github.com/cbodley/96372128ea08e4c6c1f5479f887fc5a0

while i could pass in the account-id instead of the tenant to fix that case, cross-account access is a bigger issue. with accounts, the cross-account policy evaluation requires an identity policy to allow access to the resource, and a resource policy to allow access to the requesting identity. there it's crucial that identity policy be able to use a wildcard resource that covers other accounts. the resource policy part prevents that wildcard from granting access to things it shouldn't

for non-account users, we don't have those cross-account policy rules to save us, so we need that `// You can't specify resources for someone ELSE'S account` protection. but we should skip that part for account users

so i'm thinking about using `std::optional` for that tenant argument so that account users can pass in nullopt there. does that make sense @pritha-srivastava @adamemerson?

Reply:
i pushed commit [rgw/iam: Policy takes optional tenant name](https://github.com/ceph/ceph/pull/54333/commits/f27bde2c8f2ec59d5fe541ff2da4df2ebacb21c0) for this and see the test passing now